gt/main.c

```diff
@@ -944,11 +928,14 @@ prep_notify_pkt(struct ggu_notify_pkt *ggu_pkt, struct gatekeeper_if *iface)
 				ggu_pkt->buf->l2_len);
 		notify_ipv4->total_length = rte_cpu_to_be_16(
 			ggu_pkt->buf->data_len - ggu_pkt->buf->l2_len);
+		notify_pkt->l3_len = sizeof(struct rte_ipv4_hdr);
+		set_ipv4_checksum(iface, notify_pkt, notify_ipv4);
```

Comment:
When the checksum is computed in software, doesn't it include the UDP checksum which is going to be updated next? If so, we need to reverse the computation order. Even worse, when the checksum of IPv4 is computed in software, the code has to disable the hardware computation of the UDP; otherwise, we cannot compute the checksum of IPv4 in software!

This wasn't a problem when both checksums were computed in hardware.

Reply:
While disabling the UDP hardware checksumming in `check_port_offloads()` due to IPv4 checksumming being computed in software, add a proper comment to explain the why because it's not obvious.

Reply:
Also, add a comment to explain why the UDP code is coming before the IP code once the code is properly rearranged.

Reply:
The IPv4 checksum is only computed over the IPv4 header. From the RFC that defines it:

> The checksum field is the 16-bit one's complement of the one's complement sum of all 16-bit words in the header. For purposes of computing the checksum, the value of the checksum field is zero.

So I don't think we need to worry about the UDP checksum when computing the IPv4 checksum, and can treat them independently.

Reply:
Good to know that. Please add a comment to code explaining this.